src/pinyin.cpp

```diff
@@ -0,0 +1,513 @@
+    {U"zui", U"㝡㠑㭰㭰㰎䓱䘒䘹䮔䯿厜嗺嗺嘴噿嫢嶉嶊嶵晬最栬槜槯樶檇檌欈濢璻祽稡絊纗罪蕞蟕觜辠酔酨酻醉鋷錊𠲋𠾋𠿘𡙭𡡔𡽁𡽛𢈡𢊛𣖱𣩑𥍋𥳣𥳵𦆈𦈬𦏳𦙈𦸺𧎹𧻝𨢅𨻵𨿇𩚻𩣷𩲨𪋌𪓌"},
+    {U"zun", U"䔿僔噂墫壿尊嶟拵捘撙樽繜罇袸譐遵銌鐏鱒鳟鶎鷷𠟃𠱜𤮐𥊭𥞘𥢎𦢐𦨆𦪚𨱔𩯄"},
+    {U"zuo", U"㑅㘀㘴㘸㛗㝾㤰㩇㭮㵶㸲䋏䎰䔘䝫䞢䞰䟄䟭䟶䦈佐作侳做咗唑唨嘬坐岝岞左座怍捽昨柞椊祚秨稓穝筰糳繓胙苲莋葃葄蓙袏諎酢鈼鑿阼飵𠂇𠱯𠹠𡪳𡯨𡹥𢀡𢂃𢤎𣠹𣩈𣹧𤿀𥅁𥙀𥥏𥹁𥽿𦁎𦈛𦥬𦦹𧃘𧨀𧮙𧲭𨀨𨐳𨝨𨞒𩛠𪎇𪎲"},
+    {U",", U"，"},
+    {U".", U"。"},
+    {U"?", U"？"},
+    {U"!", U"！"},
+    {U"\"", U"“”"}
+};
+
+bool pinyin_match( const std::u32string &str, const std::u32string &qry )
+{
+    // we convert the data to an unordered map to lower the cost of looking up entries.
+    // O(1) instead of O(n)
+    static std::unordered_map<char32_t, std::vector<std::u32string>> indexed_pinyin_map;
+
+    //Build the indexed map if not built yet
+    if( indexed_pinyin_map.empty() ) {
+        for( auto const &entry : pinyin_data ) {
+            //entry.first = a pinyin; entry.second = all characters that have this pronounciation
+            for( int index = 0; index < entry.second.length(); index++ ) {
+                //for each character, use it as the index and add its pinyin to the indexed map
+                char32_t current_char = entry.second.at( index );
+                try {
+                    //try to de-duplicate the entry
+                    if( std::find( indexed_pinyin_map.at( current_char ).begin(),
+                                   indexed_pinyin_map.at( current_char ).end(),
+                                   entry.first ) == indexed_pinyin_map.at( current_char ).end() ) {
+                        indexed_pinyin_map.at( current_char ).push_back( entry.first );
+                    }
+                } catch( const std::out_of_range &err ) {
+                    std::vector<std::u32string> pinyin_of_this_char;
+                    pinyin_of_this_char.push_back( entry.first );
+                    indexed_pinyin_map.insert( std::make_pair( current_char, pinyin_of_this_char ) );
+                }
+            }
+        }
+    }
+
+    int combination_index = 0;                  //how many combinations have we tried
+    bool all_combinations_tested = false;
+    while( !all_combinations_tested ) {
+        std::u32string current_combination;
+        //longest pinyin is 6 letter for a character, so we pre-allocate here
+        current_combination.reserve( str.length() * 6 );
+
+        int combination = combination_index;    //a copy so the record will not be destoryed
+        int total_combination = 0;              //the total possible amount of combinations
+
+        for( int index = 0; index < str.length(); index++ ) {
+            std::vector<std::u32string> cur_char_pinyin_list;
+            try {
+                //try to find the pinyins for the current character
+                cur_char_pinyin_list = indexed_pinyin_map.at( str.at( index ) );
+            } catch( const std::out_of_range &err ) {
+                //not a known character
+                current_combination += str.at( index );
+                continue;
+            }
```

Comment:
```suggestion
        for( const uint32_t ch : str ) {
            //try to find the pinyins for the current character
            if( indexed_pinyin_map.find( ch ) == indexed_pinyin_map.end() ) {
                //not a known character
                current_combination += ch;
                continue;
            }
            const std::vector<std::u32string> &cur_char_pinyin_list = indexed_pinyin_map.at( ch );
```
Uses a reference instead of a copy of `indexed_pinyin_map.at( ch )` to improve speed.